**Univariate Analysis: Basic Python**

Use what you learned from class to complete the following tasks using Python code and your dataframe.

Read in the mpg.csv dataset. Remember to import the pandas package everytime. Make sure the new dataframe is viewable in the output.

In [11]:
import pandas as pd

df = pd.read_csv('mpg.csv')

MPG  Cylinders  Displacement  Horse_Power  Weight  Acceleration  \
0    18.0          8         307.0          130    3504          12.0   
1    15.0          8         350.0          165    3693          11.5   
2    18.0          8         318.0          150    3436          11.0   
3    16.0          8         304.0          150    3433          12.0   
4    17.0          8         302.0          140    3449          10.5   
..    ...        ...           ...          ...     ...           ...   
387  27.0          4         140.0           86    2790          15.6   
388  44.0          4          97.0           52    2130          24.6   
389  32.0          4         135.0           84    2295          11.6   
390  28.0          4         120.0           79    2625          18.6   
391  31.0          4         119.0           82    2720          19.4   

     Model_Year                       Name  
0            70  chevrolet chevelle malibu  
1            70          buick skylark 320  
2            70         plymouth satellite  
3            70              amc rebel sst  
4            70                ford torino  
..          ...                        ...  
387          82            ford mustang gl  
388          82                  vw pickup  
389          82              dodge rampage  
390          82                ford ranger  
391          82                 chevy s-10  

[392 rows x 8 columns]

Create a DataFrame that contains the following properties and statistics of all features in the MPG.csv dataset:

* Count
* Quantiles(.25, .5, .75)
* Mean
* Min
* Max
* Standard Deviation

In [41]:

numeric_df = df.select_dtypes(include=['float64', 'int64'])
# Compute statistics
count = numeric_df.count()
quantiles = numeric_df.quantile([0.25, 0.5, 0.75])
mean = numeric_df.mean()
min_values = numeric_df.min()
max_values = numeric_df.max()
std_dev = numeric_df.std()
summary_df = pd.DataFrame({
    'Count': count,
    '25th Percentile': quantiles.loc[0.25],
    '50th Percentile (Median)': quantiles.loc[0.5],
    '75th Percentile': quantiles.loc[0.75],
    'Mean': mean,
    'Min': min_values,
    'Max': max_values,
    'Std Dev': std_dev
})

print(summary_df)


              Count  25th Percentile  50th Percentile (Median)  \
MPG             392           17.000                     22.75   
Cylinders       392            4.000                      4.00   
Displacement    392          105.000                    151.00   
Horse_Power     392           75.000                     93.50   
Weight          392         2225.250                   2803.50   
Acceleration    392           13.775                     15.50   
Model_Year      392           73.000                     76.00   

              75th Percentile         Mean     Min     Max     Std Dev  
MPG                    29.000    23.445918     9.0    46.6    7.805007  
Cylinders               8.000     5.471939     3.0     8.0    1.705783  
Displacement          275.750   194.411990    68.0   455.0  104.644004  
Horse_Power           126.000   104.469388    46.0   230.0   38.491160  
Weight               3614.750  2977.584184  1613.0  5140.0  849.402560  
Acceleration           17.025    

Given the above table, do any of the variables seem to have missing values? (HINT: Check the Count and if 0 is located in any of the variables). Which variables seem to be normally distributed at a glance? (Check how far away the min and max are from the mean). Which variables are missing from your table? Answer this in the markdown section below.

Given the count of all variables being 392, it does not appear any variables have missing values. The variables that are seemingly distributed are MPG, Cylinders, Displacement, Horse_Power, Weight, and Acceleration. The only variable missing within my table is name, but since it is a string variable instead, we can leave that out when doing statistics. 

From the data in this new DataFrame, what is the skewness of MPG? 

In [44]:
mpg_skewness = df['MPG'].skew()

print("Skewness of MPG:", mpg_skewness)
#0.45709232306041025

Skewness of MPG: 0.45709232306041025


What is the standard deviation of the weight from the mpg.csv dataset? 

In [45]:
weight_std_dev = df['Weight'].std()

print("Standard Deviation of Weight:", weight_std_dev)
#849.4025600429492

Standard Deviation of Weight: 849.4025600429492


What is the mean weight for the mpg.csv dataset?

In [46]:
mean_weight = df['Weight'].mean()

print("Mean Weight:", mean_weight)
#2977.5841836734694

Mean Weight: 2977.5841836734694


*How many* unique values are there for Name?

In [47]:
unique_names_count = df['Name'].nunique()

print("Number of unique values for Name:", unique_names_count)

Number of unique values for Name: 301


From this number alone, can you tell if any of the names are repeated in the dataset? Show in the code.

In [48]:
total_names_count = df['Name'].count()
unique_names_count = df['Name'].nunique()
repeated_names_exist = total_names_count > unique_names_count

print("Total number of entries for Name:", total_names_count)
print("Number of unique values for Name:", unique_names_count)
print("Are there any repeated names?", repeated_names_exist)

Total number of entries for Name: 392
Number of unique values for Name: 301
Are there any repeated names? True


*What are* the unique values for Cylinders?

In [50]:
unique_cylinders = df['Cylinders'].unique()

print("Unique values for Cylinders:", unique_cylinders)

Unique values for Cylinders: [8 4 6 3 5]


What is the mode of Displacement?

In [52]:
displacement_mode = df['Displacement'].mode()
print("Mode of Displacement:", displacement_mode.tolist())


Mode of Displacement: [97.0]


Show the types each of the variables (fields/columns). Do they seem fine or do any seem different than expected?

In [53]:
print(df.dtypes)

MPG             float64
Cylinders         int64
Displacement    float64
Horse_Power       int64
Weight            int64
Acceleration    float64
Model_Year        int64
Name             object
dtype: object


The data types seem fine, it makes complete sense that MPG, Displacement, and Acceleration are float points, consindering it would be needed for them to be precise. I would have expected maybe horse power and weight to also be floats vs integers. Name being an object also makes sense.

Check for missing values (NA). Are there any?

In [54]:
missing_values = df.isna().sum()

print("Missing values in each column:")
print(missing_values)

Missing values in each column:
MPG             0
Cylinders       0
Displacement    0
Horse_Power     0
Weight          0
Acceleration    0
Model_Year      0
Name            0
dtype: int64


**BONUS** *(2 pts)*: Create another DataFrame that contains the following properties and statistics of all features in the MPG.csv dataset:

* Count
* Data type
* Number unique
* Number missing
* Quantiles(.25, .5, .75)
* Mean
* Median
* Mode
* Min
* Max
* Standard Deviation
* Kurtosis
* Skewness